### PR TITLE
[LogisticRegression] Support L1 and elasticnet regularization 

### PR DESCRIPTION
**Pull request by lijinf2**
This PR maps reg_param to cuml C with relation C = 1/reg_param if reg_param != 0. else 0.
And maps elasticNet_param to cuml l1_ratio with relation l1_ratio = elasticNet. 
Finally calculates cuml penalty based on the value of reg_param and elasticNet_param.

### CODE REVIEW

**Comment by lijinf2** [comment]
build

**Comment by lijinf2** [comment]
build

**Comment by lijinf2** [comment]
build

File: python/src/spark_rapids_ml/classification.py [hunk 1]
```diff
@@ -825,6 +844,22 @@ def _logistic_regression_fit(
             )
 
             def _single_fit(init_parameters: Dict[str, Any]) -> Dict[str, Any]:
+                reg_param = (
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/classification.py:847]: Looks strange to map back to 'spark' params from C and l1_ratio.  

Could you infer the correct value of 'penalty' from already converted C and l1_ratio?

File: python/src/spark_rapids_ml/classification.py [hunk 2]
```diff
@@ -1013,6 +1069,9 @@ def intercept(self) -> float:
     def coefficientMatrix(self) -> Matrix:
         """
         Model coefficients.
+        Note Spark CPU uses denseCoefficientMatrix.compressed that may return a sparse vector
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/classification.py:1072]: Can this be addressed using cpu() method?  Convert to cpu and invoke coefficientMatrix on cpu model.   Might need to extend cpu method to multi-class case.

File: python/src/spark_rapids_ml/classification.py [hunk 3]
```diff
@@ -1027,7 +1086,16 @@ def interceptVector(self) -> Vector:
         """
         Model intercept.
         """
-        return Vectors.dense(cast(list, self.intercept_))
+        nnz = np.count_nonzero(self.intercept_)
+
+        # spark returns interceptVec.compressed
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/classification.py:1091]: Same comment as above for coefficients matrix.

File: python/src/spark_rapids_ml/classification.py [hunk 4]
```diff
@@ -1042,13 +1110,19 @@ def _get_cuml_transform_func(
         n_cols = self.n_cols
         dtype = self.dtype
 
+        penalty, C, l1_ratio = LogisticRegression._reg_params_value_mapping(
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/classification.py:1113]: Why are these needed for transform?

File: python/tests/test_logistic_regression.py [hunk 5]
```diff
@@ -85,14 +85,12 @@ def assert_transform(model: LogisticRegressionModel) -> None:
         lr_regParam_zero = LogisticRegression(
             regParam=0.0,
         )
-        assert "no regularization is not supported yet" in caplog.text
+        assert not "no regularization is not supported yet" in caplog.text
```
Comment (eordentlich) [inline python/tests/test_logistic_regression.py:88]: Probably ok to remove this.

File: python/tests/test_logistic_regression.py [hunk 6]
```diff
@@ -102,29 +100,27 @@ def assert_transform(model: LogisticRegressionModel) -> None:
 
         caplog.clear()
         lr_regParam_zero.setRegParam(0.0)
-        assert "no regularization is not supported yet" in caplog.text
+        assert not "no regularization is not supported yet" in caplog.text
```
Comment (eordentlich) [inline python/tests/test_logistic_regression.py:103]: Same here.

File: python/tests/test_logistic_regression.py [hunk 7]
```diff
@@ -148,7 +145,59 @@ def test_params(tmp_path: str, caplog: LogCaptureFixture) -> None:
     expected_cuml_params.update(
         {
             "max_iter": 30,
-            "C": 2.0,  # C should be equal to 1 / regParam
+            "penalty": "l2",
+            "C": 2.0,  # C should be equal to 2.0 / regParam
```
Comment (eordentlich) [inline python/tests/test_logistic_regression.py:149]: 1.0 / regParam

File: python/tests/test_logistic_regression.py [hunk 8]
```diff
@@ -484,11 +559,314 @@ def test_lr_fit_multiple_in_single_pass(
...
+            mlor_model.predictRaw(example.features)
+            mlor_model.predictProbability(example.features)
+
+        if isinstance(mlor_model, SparkLogisticRegressionModel):
+            assert mlor_model.hasSummary
+            mlor_model.evaluate(mdf).accuracy == mlor_model.summary.accuracy
+        else:
+            assert not mlor_model.hasSummary
+            with pytest.raises(RuntimeError, match="No training summary available"):
+                mlor_model.summary
```
Comment (eordentlich) [inline python/tests/test_logistic_regression.py:734]: Were you going to add number of iterations to summary?

**Review by eordentlich — COMMENTED** [review]
Looks good.  Minor comments.   Would be good add benchmark related updates and maybe one more run of logistic to the benchmark script with an elasticnet enabling choice of params.

**Comment by lijinf2** [comment]
build

File: python/src/spark_rapids_ml/classification.py (hunk 2 above)
Comment (lijinf2) [reply python/src/spark_rapids_ml/classification.py:1067]: Spark CPU fit uses coefficientMatrix.compressed right after training to create LogisticRegressionModel. cpu() method does not call fit so seems unlikely to help. 

It it feasible to implement the coefficientMatrix.compressed function in python, after we know how to calculate the sizes of DenseVector and SparseVector. 

Created a Jira ticket 775 to track this.

File: python/src/spark_rapids_ml/classification.py (hunk 3 above)
Comment (lijinf2) [reply python/src/spark_rapids_ml/classification.py:1086]: Similarly as above, cpu() seems not able to help here, because the coefficientMatrix.compressed is used inside CPU fit function and passed to the LogisticRegressionModel constructor. Python side is not able to pass a compressed coefficientMatrix as an argument to create a model.

Comment (lijinf2) [reply python/src/spark_rapids_ml/classification.py:1086]: The current implementation followed the Spark Scala logic of the size calculation of DenseVector and SparseVector.

File: python/src/spark_rapids_ml/classification.py (hunk 4 above)
Comment (lijinf2) [reply python/src/spark_rapids_ml/classification.py]: Good catch. No needed and deleted.

File: python/tests/test_logistic_regression.py (hunk 8 above)
Comment (lijinf2) [reply python/tests/test_logistic_regression.py:730]: Seems the summary contains many things besides number of iterations. 
Moved this to the created JIRA ticket 775 for tracking.

**Review by eordentlich — APPROVED** [review]
👍
